File: app/src/main/java/com/loabletech/bladewatch/daemon/PearMux.kt
package net.bladewatch.app.daemon

import java.nio.ByteBuffer

/**
 * Stream multiplexing over ONE Pear connection (BladeWatch-rdtj.6).
 *
 * pear-end gives each peer exactly one message channel (a Protomux `pear-connection-data`
 * channel; `connection.write` / `connection.data` over the worklet IPC), but a companion needs
 * several TCP connections to the car at once -- the live-view WebSocket, ConnectRPC calls,
 * thumbnails. This is the framing that carries them. The companion implements the same thing in
 * Dart (BladeWatch-rdtj.8); change both together.
 *
 * Protomux preserves message boundaries, so a Pear message IS a frame -- no length prefix:
 *
 *     byte 0      type:  1 OPEN, 2 DATA, 3 CLOSE, 4 WINDOW
 *     bytes 1-4   stream id, u32 big-endian, chosen by whoever opens the stream
 *     bytes 5..   OPEN:   one byte, the protocol version ([VERSION])
 *                 DATA:   1..[MAX_DATA] payload bytes
 *                 CLOSE:  nothing
 *                 WINDOW: u32 big-endian credit increment, > 0
 *
 * Only the companion opens streams; the car answers OPEN by connecting to its REMOTE loopback
 * listener and never opens a stream itself. Each direction of each stream starts with
 * [INITIAL_WINDOW] bytes of credit, and the receiver sends WINDOW as it consumes: a sender never
 * has more than a window of unacknowledged data in flight. pear-end's `connection.write` ignores
 * Protomux backpressure, so this credit is the ONLY thing that stops a live video stream from
 * piling up in the worklet when the peer's network is slower than the camera.
 */
object PearMux {
    const val VERSION: Byte = 1

    const val OPEN: Byte = 1
    const val DATA: Byte = 2
    const val CLOSE: Byte = 3
    const val WINDOW: Byte = 4

    private const val HEADER = 5

    /** Payload bytes per DATA frame: 32 KiB, ~43 KiB once pear-end's IPC base64s it. */
    const val MAX_DATA = 32 * 1024

    /** Credit each side starts with, per stream and direction. ~10 Mbit/s at a 200 ms RTT. */
    const val INITIAL_WINDOW = 256 * 1024

    class Frame(val type: Byte, val stream: Int, val payload: ByteArray)

    fun open(stream: Int): ByteArray = encode(OPEN, stream, byteArrayOf(VERSION))

    fun data(stream: Int, bytes: ByteArray, offset: Int = 0, length: Int = bytes.size): ByteArray {
        require(length in 1..MAX_DATA) { "DATA payload must be 1..$MAX_DATA bytes, was $length" }
        return encode(DATA, stream, bytes.copyOfRange(offset, offset + length))
    }

    fun close(stream: Int): ByteArray = encode(CLOSE, stream, ByteArray(0))

    fun window(stream: Int, credit: Int): ByteArray {
        require(credit > 0) { "WINDOW credit must be positive" }
        return encode(WINDOW, stream, ByteBuffer.allocate(4).putInt(credit).array())
    }

    /** The credit a WINDOW frame carries. */
    fun credit(frame: Frame): Int = ByteBuffer.wrap(frame.payload).int

    /**
     * Parses one message, or null when it is not a well-formed frame. Malformed input comes from
     * the far side of the internet, so it is a value to discard, never an exception to throw.
     */
    fun decode(message: ByteArray): Frame? {
        if (message.size < HEADER) return null
        val type = message[0]
        val stream = ByteBuffer.wrap(message, 1, 4).int
        val payload = message.copyOfRange(HEADER, message.size)
        val ok = when (type) {
            OPEN -> payload.size == 1
            DATA -> payload.size in 1..MAX_DATA
            CLOSE -> payload.isEmpty()
            WINDOW -> payload.size == 4 && ByteBuffer.wrap(payload).int > 0
            else -> false
        }
        return if (ok) Frame(type, stream, payload) else null
    }

    private fun encode(type: Byte, stream: Int, payload: ByteArray): ByteArray =
        ByteBuffer.allocate(HEADER + payload.size).put(type).putInt(stream).put(payload).array()
}

File: companion/lib/transport/pear_mux.dart
import 'dart:typed_data';

/// Stream multiplexing over ONE Pear connection -- the companion's half of BladeWatch-rdtj.6's
/// `PearMux` (app/src/main/java/com/loabletech/bladewatch/daemon/PearMux.kt). Byte for byte the
/// same format; change both together. The class doc there is the spec; in short, one Pear message
/// is one frame:
///
///     byte 0      type:  1 OPEN, 2 DATA, 3 CLOSE, 4 WINDOW
///     bytes 1-4   stream id, u32 big-endian, chosen by the opener (always the companion)
///     bytes 5..   OPEN: version byte; DATA: 1..[maxData] bytes; CLOSE: nothing; WINDOW: u32 > 0
class PearMux {
  static const int version = 1;

  static const int open = 1;
  static const int data = 2;
  static const int close = 3;
  static const int window = 4;

  static const int _header = 5;
  static const int maxData = 32 * 1024;
  static const int initialWindow = 256 * 1024;

  static Uint8List openFrame(int stream) => _encode(open, stream, Uint8List.fromList([version]));

  static Uint8List dataFrame(int stream, List<int> bytes) {
    if (bytes.isEmpty || bytes.length > maxData) {
      throw ArgumentError('DATA payload must be 1..$maxData bytes, was ${bytes.length}');
    }
    return _encode(data, stream, bytes);
  }

  static Uint8List closeFrame(int stream) => _encode(close, stream, const []);

  static Uint8List windowFrame(int stream, int credit) {
    if (credit <= 0) throw ArgumentError('WINDOW credit must be positive');
    return _encode(window, stream, (ByteData(4)..setUint32(0, credit)).buffer.asUint8List());
  }

  /// One message, or null when it is not a well-formed frame. It comes from the far side of the
  /// internet: a value to discard, never an exception to throw.
  static MuxFrame? decode(Uint8List message) {
    if (message.length < _header) return null;
    final type = message[0];
    final stream = ByteData.sublistView(message, 1, 5).getUint32(0);
    final payload = Uint8List.sublistView(message, _header);
    final ok = switch (type) {
      open => payload.length == 1,
      data => payload.isNotEmpty && payload.length <= maxData,
      close => payload.isEmpty,
      window => payload.length == 4 && ByteData.sublistView(payload).getUint32(0) > 0,
      _ => false,
    };
    return ok ? MuxFrame(type, stream, payload) : null;
  }

  static Uint8List _encode(int type, int stream, List<int> payload) {
    final out = Uint8List(_header + payload.length);
    out[0] = type;
    ByteData.sublistView(out, 1, 5).setUint32(0, stream);
    out.setRange(_header, out.length, payload);
    return out;
  }
}

class MuxFrame {
  final int type;
  final int stream;
  final Uint8List payload;

  const MuxFrame(this.type, this.stream, this.payload);

  /// The credit a WINDOW frame carries.
  int get credit => ByteData.sublistView(payload).getUint32(0);
}

/// What the companion needs from a Pear connection to the car, and nothing more: messages in,
/// messages out. The real one wraps flutter_pear's `PearConnection`; tests use an in-memory pair.
abstract class PeerLink {
  /// Each event is one Pear message (Protomux keeps boundaries): exactly one [PearMux] frame.
  Stream<Uint8List> get messages;

  Future<void> send(Uint8List message);
}
